File: leetcode_py/Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# note that if judged too late, i.e. at the leaf, the answer will be doubled.

class Solution(object):
    def helper(self, cur, path, res):
        newpath = path + str(cur.val) + '->'
        if not cur.left and not cur.right:
            if len(newpath) > 2:
                res.append(newpath[:-2])
        else:
            if cur.left:
                self.helper(cur.left, newpath, res)
            if cur.right:
                self.helper(cur.right, newpath, res)

    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        if not root:
            return []
        res = []
        self.helper(root, '', res)
        return res

File: leetcode_js/Binary Tree Paths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */

var helper = function(root, path, res){
    if(!root) return;
    var tmppath = path+root.val+'->';
    if(!root.left && !root.right){
        res.push(tmppath.length>0?tmppath.substr(0, tmppath.length-2):'');
        return;
    }
    if(root.left){
        helper(root.left, tmppath, res);
    }
    if(root.right){
        helper(root.right, tmppath, res);
    }
}
var binaryTreePaths = function(root) {
    var res = [];
    helper(root, '', res);
    return res;
};
root = new TreeNode(1)
root.left = new TreeNode(2);
console.log(binaryTreePaths(root))